xrootd/encoder/encoder.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package encoder // import "go-hep.org/x/hep/xrootd/encoder"
+
+import (
+	"encoding/binary"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// MarshalRequest encodes request body alongside with request and stream ids
+func MarshalRequest(requestID uint16, streamID protocol.StreamID, requestBody interface{}) ([]byte, error) {
+	requestHeader := make([]byte, 4)
```

Comment:
Sorry, I'm not sure that I totally understand. Are you suggesting to preallocate an array for the whole request and not just header?

Reply:
Yes, it would be great to have something like:

```go
req := make([]byte, 4, n)
```

Where n isn't too big to not preallocate too much but big enough so most calls to MarshalRequest wouldn't need to copy the whole buffer over in the call to append.

This could be marked as a TODO :)